File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        int hasCycle = 0;
        ListNode * slow = head;
        ListNode *fast = head;
        while(fast && fast->next)
        {
            slow = slow->next;
            fast = fast->next->next;
            if(slow == fast)
            {
                hasCycle=1;
                break;
            }

        }

        if(hasCycle==0)
        {
            return NULL;
        }
        ListNode *ptr = head;
        while(ptr!=slow)
        {
            ptr = ptr->next;
            slow = slow->next;
        }
        return ptr;
        
    }
};

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None 
        slow =head 
        fast = head 
        while fast and fast.next:
            slow = slow.next 
            fast = fast.next.next 
            if slow == fast:
                new = head 
                while new and slow:
                    if new ==slow:
                        return new
                    new = new.next 
                    slow = slow.next 
                    
        return None
        
        

        